### PR TITLE
LA-191: Get all functionality Domain + Data layer

### PR DESCRIPTION
**Pull request by manhduc-nguyen**

### CODE REVIEW

**Comment by hoangdat** [comment]
LGTM, go ahead 💪 💪 💪

File: lib/presentation/redux/reducers/functionality_reducer.dart [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import 'package:linshare_flutter_app/presentation/redux/states/functionality_state.dart';
+import 'package:redux/redux.dart';
+
+final functionalityReducer = combineReducers<FunctionalityState>([
+  TypedReducer<FunctionalityState, StartFunctionalityLoadingAction>((FunctionalityState state, _) => state.startLoadingState()),
+  TypedReducer<FunctionalityState, FunctionalityAction>((FunctionalityState state, FunctionalityAction action) => state.sendViewState(viewState: action.viewState)),
+  TypedReducer<FunctionalityState, FunctionalityGetAllAction>((FunctionalityState state, FunctionalityGetAllAction action) =>
+      state.setFunctionalityList(
+          action.viewState.fold(
+                  (failure) => [],
```
Comment (hoangdat) [inline lib/presentation/redux/reducers/functionality_reducer.dart:44]: how about when we have no list of functionalities?

Comment (manhduc-nguyen) [reply lib/presentation/redux/reducers/functionality_reducer.dart:44]: as you see, I will return an empty array if we don't have data from api. So it does not affect our app.

Comment (hoangdat) [reply lib/presentation/redux/reducers/functionality_reducer.dart:44]: but, the after that life still goes on.
1. How about: we have no list but admin still disable workgroup?
2. ...to be continue :)

Comment (manhduc-nguyen) [reply lib/presentation/redux/reducers/functionality_reducer.dart:44]: Oh if this is a `must have` before start app. I will change this.

File: lib/presentation/redux/states/functionality_state.dart [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  @override
+  FunctionalityState startLoadingState() {
+    return FunctionalityState(Right(LoadingState()), functionalityList);
+  }
+}
+
+extension FunctionalityStateExtension on FunctionalityState {
+  bool isSharedSpaceEnable() {
```
Comment (imGok) [inline lib/presentation/redux/states/functionality_state.dart:71]: Why not `bool isFunctionnalityEnable(FunctionalityIdentifier fonc)` ?

**Review by imGok — APPROVED** [review]
LGTM, just a note : plurial of **functionality** is **functionalities**

**Comment by manhduc-nguyen** [comment]
> LGTM, just a note : plurial of **functionality** is **functionalities**

yeah I know, but Android studio alerts me a typo warning if I use `functionalities` and it is uncomfortable for me to see that lol

File: lib/presentation/redux/states/functionality_state.dart (hunk 2 above)
Comment (manhduc-nguyen) [reply lib/presentation/redux/states/functionality_state.dart]: @imGok done

**Comment by imGok** [comment]
> yeah I know, but Android studio alerts me a typo warning if I use `functionalities` and it is uncomfortable for me to see that lol

sure no problem to me

**Review by imGok — APPROVED** [review]

File: domain/lib/src/model/functionality/functionality_identifier.dart [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+//
+
+enum FunctionalityIdentifier {
+  UNKNOWN_IDENTIFIER,
+  WORK_GROUP,
+  CONTACTS_LIST__CREATION_RIGHT,
+  UPLOAD_REQUEST__PROLONGATION,
```
Comment (hoangdat) [inline domain/lib/src/model/functionality/functionality_identifier.dart:37]: why we need these?

Comment (manhduc-nguyen) [reply domain/lib/src/model/functionality/functionality_identifier.dart]: done, pls check again

**Review by hoangdat — APPROVED** [review]

**Pull request merged**